libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -38,8 +52,152 @@ class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value
   counting_semaphore& operator=(const counting_semaphore&) = delete;
 };
 
+template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
+struct fair_counting_semaphore
+{
+};
+
 template <thread_scope _Sco>
-using binary_semaphore = counting_semaphore<_Sco, 1>;
+struct binary_semaphore : fair_counting_semaphore<_Sco, 1>
+{
+public:
+    _LIBCUDACXX_INLINE_VISIBILITY static constexpr ptrdiff_t max() noexcept
+    {
+        return static_cast<ptrdiff_t>(1);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY bool try_acquire()
+    {
+        ptrdiff_t ctr = counter.load(cuda::memory_order_relaxed);
+        return try_acq_impl(ctr);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY void acquire()
+    {
+        ptrdiff_t ctr = counter.load(cuda::memory_order_relaxed);
+        while (!try_acq_impl(ctr))
+        {
+            counter.wait(ctr);
+        }
+        int32_t cur_tickets = (ticketing.load(cuda::memory_order_relaxed) & 0xFFFFFFFF);
+        auto limit = _CUDA_VSTD::numeric_limits<int32_t>::max();
+        if (limit >= cur_tickets + 1)
+        {
+            auto prev = ticketing.fetch_add(1);
+            while (!our_turn(prev + 1))
+            {
+                simple_backoff();
+            }
+        }
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY bool __acquire_slow_timed(
+        _CUDA_VSTD::chrono::nanoseconds const &__rel_time)
+    {
+        return __libcpp_thread_poll_with_backoff(
+            [this]()
+            {
+                ptrdiff_t ctr = counter.load(cuda::memory_order_relaxed);
+                while (!try_acq_impl(ctr))
+                {
+                    counter.wait(ctr);
+                }
+                int32_t cur_tickets = (ticketing.load(cuda::memory_order_relaxed) & 0xFFFFFFFF);
+                auto limit = _CUDA_VSTD::numeric_limits<int32_t>::max();
+                if (limit >= cur_tickets + 1)
+                {
+                    auto prev = ticketing.fetch_add(1);
+                    while (!our_turn(prev + 1))
+                    {
+                        simple_backoff();
+                    }
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            },
+            __rel_time);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY void release(ptrdiff_t update = 1)
+    {
+        counter.fetch_add(1);
+        ticketing.fetch_add(static_cast<ptrdiff_t>(1) << 32, cuda::memory_order_relaxed);
+        // we need to notify all because no guarantee of who
+        // gets it
+        counter.notify_all();
+        ticketing.notify_all();
```

Comment:
Like I've outlined in my PR summary I am not too fond of using `notify_all()` in general because as I understand it we don't have a way of guaranteeing that threads are notified in a FIFO order. So to avoid hanging, we need to wake everyone up. I just put this in for now but I think more design considerations are required here.

Reply:
I'm not immediately aware of an alternative design since `notify[_all]` is a noop. CUDA threads can never be explicitly yielded. The best we can do on CUDA is `nanosleep(N)` and recheck.

Reply:
> I'm not immediately aware of an alternative design since `notify[_all]` is a noop. CUDA threads can never be explicitly yielded. The best we can do on CUDA is `nanosleep(N)` and recheck.

I might be asking a stupid question so forgive if I am but shold this semaphore be useable from both host and device or only device? I made it with the assumption it works on both